wknml/nml_utils.py

```diff
@@ -0,0 +1,104 @@
+from typing import Union, Dict, List
+from math import sqrt, ceil
+from copy import deepcopy
+import networkx as nx
+import numpy as np
+
+from . import NML
+from .nml_generation import generate_graph, generate_nml
+
+
+def ensure_max_edge_length(nml_or_graph: Union[NML, nx.Graph], max_length: int) -> Union[NML, nx.Graph]:
+    # it is easier to operate on a graph
+    if isinstance(nml_or_graph, nx.Graph):
+        nml_graph = nml_or_graph
+    else:
+        nml_graph, parameter_dict = generate_graph(nml_or_graph)
+    max_id = detect_max_node_id_from_all_graphs(nml_graph)
+    next_valid_id = max_id + 1
+    for group in nml_graph.values():
+        for graph in group:
+            next_valid_id = ensure_max_edge_length_in_tree(graph, max_length, next_valid_id)
+
+    # return the same format as the input
+    if isinstance(nml_or_graph, nx.Graph):
+        return nml_graph
+    else:
+        return generate_nml(nml_graph, parameter_dict, globalize=False)
+
+
+def calculate_distance_between_nodes(node1: Dict, node2: Dict) -> float:
+    node1_position = node1['position']
+    node2_position = node2['position']
+    return sqrt((node1_position[0] - node2_position[0]) ** 2 +
+                (node1_position[1] - node2_position[1]) ** 2 +
+                (node1_position[2] - node2_position[2]) ** 2)
+
+
+def get_vector_between_nodes(node1: Dict, node2: Dict) -> List[int]:
+    node1_position = node1['position']
+    node2_position = node2['position']
+    return [node2_position[0] - node1_position[0],
+            node2_position[1] - node1_position[1],
+            node2_position[2] - node1_position[2]]
+
+
+def get_padding_node_position(node1: Dict, node2: Dict, relative_distance_along_vector: float) -> List[int]:
+    node1_position = node1['position']
+    vector_between_nodes = get_vector_between_nodes(node1, node2)
+    return [node1_position[0] + vector_between_nodes[0] * relative_distance_along_vector,
```

Comment:
If you turn your vectors into numpy arrays things become easier.
```suggestion
    return node1_position + vector_between_nodes * relative_distance_along_vector
```